File: deploy_remote_config.js
#!/usr/bin/env node
/**
 * Firebase Remote Config Deployment Script (Node.js)
 * Amazon Rewards ve Points sistemi için Remote Config parametrelerini yükler
 */

const admin = require('firebase-admin');
const fs = require('fs');
const path = require('path');

// Firebase Admin SDK başlat
async function initializeFirebase() {
  try {
    // Service account key dosyası var mı kontrol et
    const serviceAccountPath = path.join(__dirname, 'firebase-service-account.json');
    
    if (fs.existsSync(serviceAccountPath)) {
      const serviceAccount = require(serviceAccountPath);
      admin.initializeApp({
        credential: admin.credential.cert(serviceAccount),
      });
      console.log('✅ Firebase Admin SDK başlatıldı (Service Account)');
    } else {
      // Application Default Credentials kullan (gcloud veya Firebase CLI ile login)
      admin.initializeApp();
      console.log('✅ Firebase Admin SDK başlatıldı (Application Default Credentials)');
    }
    
    return true;
  } catch (error) {
    console.error('❌ Firebase Admin SDK başlatılamadı:', error.message);
    console.log('\n💡 Çözüm:');
    console.log('   1. Firebase CLI ile login: firebase login');
    console.log('   2. Veya service account key ekle: firebase-service-account.json');
    return false;
  }
}

// Remote Config'i yükle
async function deployRemoteConfig(configPath) {
  try {
    // Config dosyasını oku
    const configJson = fs.readFileSync(configPath, 'utf8');
    const config = JSON.parse(configJson);
    
    // Remote Config template oluştur
    const remoteConfig = admin.remoteConfig();
    const template = remoteConfig.createTemplateFromJSON(configJson);
    
    // Template'i yükle
    console.log('📤 Remote Config yükleniyor...');
    const updatedTemplate = await remoteConfig.publishTemplate(template);
    
    console.log('✅ Remote Config başarıyla yüklendi!');
    console.log(`   Version: ${updatedTemplate.version.versionNumber}`);
    console.log(`   Update Time: ${updatedTemplate.version.updateTime}`);
    
    return true;
  } catch (error) {
    console.error('❌ Remote Config yüklenemedi:', error.message);
    
    if (error.code === 'permission-denied') {
      console.log('\n💡 Çözüm:');
      console.log('   Firebase projesine yazma yetkisine sahip olduğunuzdan emin olun');
      console.log('   firebase login --reauth');
    }
    
    return false;
  }
}

// Ana fonksiyon
async function main() {
  console.log('🚀 Firebase Remote Config Deployment (Node.js)');
  console.log('='.repeat(50));
  console.log();
  
  // Config dosyası yolu
  const configPath = path.join(__dirname, 'remote_config_merged.json');
  
  if (!fs.existsSync(configPath)) {
    console.error(`❌ Config dosyası bulunamadı: ${configPath}`);
    console.log('   Önce deploy_remote_config.py scriptini çalıştırın');
    process.exit(1);
  }
  
  // Firebase'i başlat
  const initialized = await initializeFirebase();
  if (!initialized) {
    process.exit(1);
  }
  
  console.log();
  
  // Remote Config'i yükle
  const deployed = await deployRemoteConfig(configPath);
  
  if (deployed) {
    console.log();
    console.log('='.repeat(50));
    console.log('🎉 Tamamlandı!');
    console.log();
    console.log('📋 Sonraki Adımlar:');
    console.log('   1. Firebase Console → Remote Config');
    console.log('   2. Yeni parametreleri kontrol edin');
    console.log('   3. Değişiklikler 1 saat içinde uygulamaya yansıyacak');
  } else {
    process.exit(1);
  }
}

// Script'i çalıştır
main().catch((error) => {
  console.error('❌ Beklenmeyen hata:', error);
  process.exit(1);
});

File: deploy_remote_config.py
#!/usr/bin/env python3
"""
Firebase Remote Config Deployment Script
Amazon Rewards ve Points sistemi için Remote Config parametrelerini yükler
"""

import json
import sys
import subprocess
import os
from pathlib import Path

def get_firebase_access_token():
    """Firebase access token'ı al"""
    try:
        # Firebase CLI ile access token al
        result = subprocess.run(
            ['firebase', 'login:ci'],
            capture_output=True,
            text=True,
            check=False
        )
        
        # Alternatif: firebase use ile project ID al, sonra gcloud ile token al
        # Ya da kullanıcıdan manuel token iste
        print("⚠️  Firebase access token gerekiyor")
        print("   İki yöntem:")
        print("   1. Firebase Console → Project Settings → Service Accounts → Generate New Private Key")
        print("   2. Veya firebase login:ci komutunu çalıştırın")
        return None
    except Exception as e:
        print(f"❌ Token alma hatası: {e}")
        return None

def merge_remote_config(current_config_path, new_config_path, output_path):
    """Mevcut ve yeni Remote Config'i birleştir"""
    # Mevcut config'i yükle
    if os.path.exists(current_config_path):
        with open(current_config_path, 'r', encoding='utf-8') as f:
            current_config = json.load(f)
    else:
        current_config = {"parameters": {}}
    
    # Yeni config'i yükle
    with open(new_config_path, 'r', encoding='utf-8') as f:
        new_config = json.load(f)
    
    # Parametreleri birleştir (yeni parametreler mevcutları günceller)
    merged_parameters = current_config.get("parameters", {}).copy()
    merged_parameters.update(new_config.get("parameters", {}))
    
    # Birleştirilmiş config
    merged_config = {
        "parameters": merged_parameters,
        "version": new_config.get("version", {
            "versionNumber": "1",
            "updateTime": "2025-01-20T00:00:00Z",
            "updateUser": {
                "email": "qanta@remote-config.com"
            },
            "description": "Amazon Rewards ve Points sistemi için Remote Config ayarları",
            "updateOrigin": "REST_API",
            "updateType": "INCREMENTAL_UPDATE"
        })
    }
    
    # Kaydet
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(merged_config, f, indent=2, ensure_ascii=False)
    
    print(f"✅ Birleştirilmiş config kaydedildi: {output_path}")
    return merged_config

def main():
    print("🚀 Firebase Remote Config Deployment")
    print("=" * 50)
    print()
    
    # Dosya yolları
    script_dir = Path(__file__).parent
    current_config_path = script_dir / "remote_config_current.json"
    new_config_path = script_dir / "firebase_remote_config_amazon_rewards_points.json"
    merged_config_path = script_dir / "remote_config_merged.json"
    
    # Dosya kontrolü
    if not new_config_path.exists():
        print(f"❌ Yeni config dosyası bulunamadı: {new_config_path}")
        sys.exit(1)
    
    # Mevcut config'i çek
    print("📥 Mevcut Remote Config çekiliyor...")
    try:
        subprocess.run(
            ['firebase', 'remoteconfig:get', '-o', str(current_config_path)],
            check=True,
            capture_output=True
        )
        print("✅ Mevcut config çekildi")
    except subprocess.CalledProcessError:
        print("⚠️  Mevcut config çekilemedi (devam ediliyor...)")
        current_config_path = None
    
    # Config'leri birleştir
    print()
    print("🔀 Config'ler birleştiriliyor...")
    merged_config = merge_remote_config(
        current_config_path if current_config_path and current_config_path.exists() else None,
        new_config_path,
        merged_config_path
    )
    
    # Parametre sayısı
    param_count = len(merged_config.get("parameters", {}))
    print(f"📊 Toplam parametre sayısı: {param_count}")
    print()
    
    # Yeni parametreleri listele
    if current_config_path and current_config_path.exists():
        with open(current_config_path, 'r', encoding='utf-8') as f:
            current_params = set(json.load(f).get("parameters", {}).keys())
    else:
        current_params = set()
    
    new_params = set(merged_config.get("parameters", {}).keys())
    added_params = new_params - current_params
    
    if added_params:
        print("🆕 Eklenen yeni parametreler:")
        for param in sorted(added_params):
            print(f"   - {param}")
        print()
    
    # Firebase REST API ile yükleme talimatları
    print("=" * 50)
    print("📤 Remote Config'i yüklemek için:")
    print()
    print("Yöntem 1: Firebase Console (Önerilen)")
    print("   1. Firebase Console → Remote Config")
    print("   2. 'remote_config_merged.json' dosyasını aç")
    print("   3. Her parametreyi manuel olarak ekle veya")
    print("   4. Firebase Console → Remote Config → '...' → 'Import from file'")
    print()
    print("Yöntem 2: Firebase REST API")
    print("   curl -X PUT https://firebaseremoteconfig.googleapis.com/v1/projects/YOUR_PROJECT_ID/remoteConfig \\")
    print("     -H 'Authorization: Bearer YOUR_ACCESS_TOKEN' \\")
    print("     -H 'Content-Type: application/json' \\")
    print("     -d @remote_config_merged.json")
    print()
    print("Yöntem 3: Firebase CLI (manuel)")
    print("   Firebase CLI'da Remote Config set komutu yok.")
    print("   Ancak mevcut config'i çekip, yeni parametreleri ekleyip")
    print("   Firebase Console'dan import edebilirsiniz.")
    print()
    print(f"✅ Birleştirilmiş config hazır: {merged_config_path}")
    print()

if __name__ == "__main__":
    main()

